modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckDistributedProcess.java

```diff
@@ -0,0 +1,549 @@
+
+        return FINISHED_FUT;
+    }
+
+    /** Phase 2 beginning.  */
+    private IgniteInternalFuture<HashMap<PartitionKeyV2, PartitionHashRecordV2>> validateParts(
+        SnapshotCheckProcessRequest incReq) {
+        SnapshotCheckProcessRequest locReq;
+
+        if (stopAndCleanOnError(incReq, null) || (locReq = requests.get(incReq.snapshotName())) == null)
+            return FINISHED_FUT;
+
+        assert locReq.equals(incReq);
+
+        // Store metas to collect cluster operation result laster.
+        GridFutureAdapter<SnapshotPartitionsVerifyTaskResult> clusterOpFut = clusterOpFuts.get(incReq.requestId());
+
+        if (clusterOpFut != null)
+            locReq.metas = incReq.metas;
+
+        // Local meta might be null if current node started after the snapshot creation or placement.
+        if (!incReq.nodes.contains(kctx.localNodeId()) || locReq.meta() == null)
+            return FINISHED_FUT;
+
+        GridFutureAdapter<HashMap<PartitionKeyV2, PartitionHashRecordV2>> locPartsChkFut = new GridFutureAdapter<>();
+
+        locReq.fut(locPartsChkFut);
+
+        ExecutorService executor = kctx.cache().context().snapshotMgr().snapshotExecutorService();
+
+        executor.submit(() -> stopFutureOnAnyFailure(locPartsChkFut, () -> {
+            // An error can occure when the local future is still null.
+            if (locReq.error() != null)
+                locPartsChkFut.onDone(locReq.error());
+
+            if (locPartsChkFut.isDone())
+                return;
+
+            File snpDir = kctx.cache().context().snapshotMgr().snapshotLocalDir(locReq.snapshotName(), locReq.snapshotPath());
+
+            try {
+                Map<PartitionKeyV2, PartitionHashRecordV2> res = kctx.cache().context().snapshotMgr().checker()
+                    .checkPartitions(locReq.meta(), snpDir, locReq.groups(), false, true, false);
+
+                locPartsChkFut.onDone(res instanceof HashMap ? (HashMap<PartitionKeyV2, PartitionHashRecordV2>)res
+                    : new HashMap<>(res));
+            }
+            catch (IgniteCheckedException e) {
+                throw new IgniteException("Failed to calculate snapshot partition hashes, req: " + incReq, e);
+            }
+
+            // No need to wait to the cealup if current node is just a worker.
+            if (!kctx.localNodeId().equals(locReq.opCoordId) && clusterOpFut == null)
+                clean(locReq.reqId, null, null, null);
+        }));
+
+        return locPartsChkFut;
+    }
+
+    /**
+     * If required, stops and clean related validation if errors occured
+     *
+     * @return {@code True} if the validation stopped and cleaned. {@code False} otherwise.
+     */
+    private boolean stopAndCleanOnError(SnapshotCheckProcessRequest req,
+        @Nullable Map<UUID, Throwable> occuredErrors) {
+        assert req != null;
+
+        if (!F.isEmpty(occuredErrors)) {
+            occuredErrors = occuredErrors.entrySet().stream()
```

Comment:
I suppose the errors can appear only from nodes that are part of the request? Because these errors are passed in reduce with the request

Reply:
Well, it should. But the process works on any node. Theoretically, any unrelated, unexpected exception might occur on any node. Let's keep an assert.